safenode/src/transfers/mod.rs

```diff
@@ -0,0 +1,216 @@
+    protocol::types::{
+        address::DbcAddress,
+        fees::{FeeCiphers, RequiredFee, SpendPriority, SpendQ},
+    },
+    storage::SpendStorage,
+};
+
+use sn_dbc::{DbcId, DbcTransaction, MainKey, SignedSpend, Token};
+
+use std::collections::{BTreeMap, BTreeSet};
+use xor_name::XorName;
+
+const STARTING_FEE: u64 = 4000; // 0.000004 SNT
+
+pub(super) struct Transfers {
+    our_name: XorName,
+    node_reward_key: MainKey,
+    spend_queue: SpendQ<SignedSpend>,
+    storage: SpendStorage,
+}
+
+impl Transfers {
+    /// Create a new instance of `Transfers`.
+    pub(crate) fn new(our_name: XorName, node_reward_key: MainKey) -> Self {
+        Self {
+            our_name,
+            node_reward_key,
+            spend_queue: SpendQ::with_fee(STARTING_FEE),
+            storage: SpendStorage::new(),
+        }
+    }
+
+    /// Get Spend from local store.
+    pub(crate) async fn get(&self, address: DbcAddress) -> Result<SignedSpend> {
+        self.storage.get(address).await
+    }
+
+    /// Get the required fee for the specified spend priority.
+    pub(crate) fn get_required_fee(&self, dbc_id: DbcId, priority: SpendPriority) -> RequiredFee {
+        let amount = self.current_fee(priority);
+
+        debug!("Returned amount for priority {priority:?}: {amount}");
+
+        RequiredFee::new(Token::from_nano(amount), dbc_id, &self.node_reward_key)
+    }
+
+    /// Get the current fee for the specified spend priority.
+    fn current_fee(&self, priority: SpendPriority) -> u64 {
+        let spend_q_snapshot = self.spend_queue.snapshot();
+        let spend_q_stats = spend_q_snapshot.stats();
+        spend_q_stats.map_to_fee(priority)
+    }
+
+    /// Tries to add a double spend that was detected by the network.
+    pub(crate) async fn try_add_double(
+        &self,
+        a_spend: &SignedSpend,
+        b_spend: &SignedSpend,
+    ) -> Result<()> {
+        self.storage.try_add_double(a_spend, b_spend).await
+    }
+
+    /// Tries to add a new spend to the queue.
+    ///
+    /// All the provided data will be validated, and
+    /// if it is valid, the spend will be pushed onto the queue.
+    pub(crate) async fn try_add(
+        &mut self,
+        signed_spend: Box<SignedSpend>,
+        source_tx: Box<DbcTransaction>,
+        fee_ciphers: BTreeMap<XorName, FeeCiphers>,
+        parent_spends: BTreeSet<SignedSpend>,
+    ) -> Result<()> {
+        // 1. Validate the tx hash.
+        // Ensure that the provided src tx is the same as the
+        // one we have the hash of in the signed spend.
+        let provided_src_tx_hash = source_tx.hash();
+        let signed_src_tx_hash = signed_spend.src_tx_hash();
+
+        if provided_src_tx_hash != signed_src_tx_hash {
+            return Err(Error::SignedSrcTxHashDoesNotMatchProvidedSrcTxHash {
+                signed_src_tx_hash,
+                provided_src_tx_hash,
+            });
+        }
+
+        // 2. Try extract the fee paid for this spend, and validate it.
```

Comment:
Minor style issue. If the code says `validate_fee`, the comment isn't providing much in addition. Personally, if you wanted to describe the step by step process, I think that would be better done in the doc comments and let the code here speak for itself.

Reply:
Could be. But I go quite heavily on explaining step by step. I prefer that over assuming that code is self-explaining. Most people assume to much with that, as the code is known to them, with tons of implicit knowledge.